### PR TITLE
[clang][driver] Don't warn when -S and -c are used together without -fsyntax-only

### LINKED ISSUE #104347: [clang][driver] -S and -c are incompatible after #101103 when using Werror, even without -fsyntax-only
After https://github.com/llvm/llvm-project/pull/101103, `-S` and `-c` options seem to be incompatible when using -Werror. This doesn't seem to be intentional, from what I can tell.

The following comands result in `clang: error: argument unused during compilation: '-c' [-Werror,-Wunused-command-line-argument]`
```
touch foo.c
~/llvm-project/build/bin/clang -S -O0 -Werror  -c .foo.c -o foo.c.o
```

CC: @MaskRay

### PR DESCRIPTION
**Pull request by ilovepi**
Claim `-c` when using `-S`, so that we don't warn for unused arguments
when they are used together without `-fsyntax-only`.

Fixes #104347

### CODE REVIEW

File: clang/test/Driver/warn-fsyntax-only.c [hunk 1]
```diff
@@ -1,7 +1,14 @@
 // RUN: %clang --target=x86_64 -fsyntax-only -E %s 2>&1 | FileCheck %s --check-prefix=CHECK-PP
 // RUN: %clang --target=x86_64 -fsyntax-only -S %s 2>&1 | FileCheck %s --check-prefix=CHECK-ASM
 // RUN: %clang --target=x86_64 -fsyntax-only -c %s 2>&1 | FileCheck %s --check-prefix=CHECK-OBJ
+// RUN: %clang --target=x86_64 -fsyntax-only -S -c %s 2>&1 | FileCheck %s --check-prefixes=CHECK-ASM,CHECK-OBJ
```
Comment (ilovepi) [inline clang/test/Driver/warn-fsyntax-only.c:4]: Do these need the target set? Seems like this should be fine for any target, right?

File: clang/lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -377,8 +377,10 @@ phases::ID Driver::getFinalPhase(const DerivedArgList &DAL,
   // -S only runs up to the backend.
   } else if ((PhaseArg = DAL.getLastArg(options::OPT_S))) {
     FinalPhase = phases::Backend;
+    // Avoid unused argument warnings when combining -S and -c.
+    DAL.ClaimAllArgs(options::OPT_c);
 
-  // -c compilation only runs up to the assembler.
+    // -c compilation only runs up to the assembler.
```
Comment (ilovepi) [inline clang/lib/Driver/Driver.cpp:383]: Not sure why `git clang-format` wanted to update this comment, but 🤷 .

**Comment by MaskRay** [comment]
> Fixes https://github.com/llvm/llvm-project/issues/104347

I made a comment on #104347  . The warning for `-S -c` works as intended.

**Review by MaskRay — CHANGES_REQUESTED** [review]
.

**Pull request closed without merging**